<a href="https://colab.research.google.com/github/olkasadova/data-and-python/blob/main/Data_Analysis_Worksheets/Sorting_and_cleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Sorting and cleaning
---



In order to effectively analyse a dataset, often we need to prepare it first.
Before a dataset is ready to be analysed we might need to:  

* sort the data (can be a series or dataframe)  
* remove any NaN values or drop NA values   
* remove duplicate records (identical rows)  
* normalise data in dataframe columns so that has a common scale [reference](https://towardsai.net/p/data-science/how-when-and-why-should-you-normalize-standardize-rescale-your-data-3f083def38ff#:~:text=Similarly%2C%20the%20goal%20of%20normalization,dataset%20does%20not%20require%20normalization.&text=So%20we%20normalize%20the%20data,variables%20to%20the%20same%20range.)

## Sorting the data  
---


Typically we want to sort data by the values in one or more columns in the dataframe  

To sort the dataframe by series we use the pandas function **sort_values()**.  

By default `sort_values()` sorts into ascending order.

* sort by a single column e.g.
  * `df.sort_values("Make") `
* sort by multiple columns e.g.
  * `df.sort_values(by = ["Model", "Make"]) `
    * this sorts by Model, then my Make
* sort in *descending* order
  * `df.sort_values(by = "Make", ascending = False)`
  * `df.sort_values(by = ["Make", "Model"], ascending = False])`  

Dataframes are mostly immutable, changes like sort_values do not change the dataframe permanently, they just change it for the time that the instruction is being used.

`df.sort_values(by='Make')` *dataframe is now in sorted order and can be copied to a new dataframe*  
`df` *original dataframe, df, will be as it was - unsorted*

To split the dataframe after sorting, do this in the same instruction, e.g.:

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]]`

This sorts on Make and then Model in descending order, then splits off the Make and Model columns.

`df.sort_values(by = ["Make", "Model"], ascending = False])[["Make", "Model"]].head()`

This sorts on Make and then Model, then splits off the Make and Model columns and then splits off the first 5 rows.

### Exercise 1 - get data, sort by happiness score
---

Read data from the Excel file on Happiness Data at this link: https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true

Display first 5 rows of data  

The data is currently sorted by Happiness Rank...
*  sort the data by Happiness Score in ascending order
*  display sorted table

**Test output**:  
The lowest score (displayed first) is 2.839, Togo  
The highest score (displayed last) is 7.587, Switzerland  



In [19]:
import pandas as pd
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness = pd.read_excel(url)

sorted_list_Acs= happiness.sort_values("Happiness Score").head(5)
print ("The lowest score is: ", sorted_list_Acs ["Country"], ["Happiness Score"])

sorted_list_Des= happiness.sort_values(by= ["Happiness Score"], ascending = False).head(5)
print ("The highest score is: ", sorted_list_Des ["Country"], ["Happiness Score"])

The lowest score is:  157       Togo
156    Burundi
155      Syria
154      Benin
153     Rwanda
Name: Country, dtype: object ['Happiness Score']
The highest score is:  0    Switzerland
1        Iceland
2        Denmark
3         Norway
4         Canada
Name: Country, dtype: object ['Happiness Score']


### Exercise 2 - sort by multiple columns, display the first 5 rows
---

1. sort the data by Economy (GDP per Capita) and Health (Life Expectancy) in ascending order
2. display the first 5 rows of sorted data

**Test output**:  
Records 122, 127, 147, 100, 96

In [30]:
import pandas as pd
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness = pd.read_excel(url)

sorted_Economy = happiness.sort_values (by= [ "Health (Life Expectancy)", "Economy (GDP per Capita)"],ascending = True).head(5)
print (sorted_Economy )

                      Country              Region  Happiness Rank  \
122              Sierra Leone  Sub-Saharan Africa             123   
127                  Botswana  Sub-Saharan Africa             128   
147  Central African Republic  Sub-Saharan Africa             148   
100                 Swaziland  Sub-Saharan Africa             101   
96                    Lesotho  Sub-Saharan Africa              97   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
122            4.507         0.07068                   0.33024  0.95571   
127            4.332         0.04934                   0.99355  1.10464   
147            3.678         0.06112                   0.07850  0.00000   
100            4.867         0.08742                   0.71206  1.07284   
96             4.898         0.09438                   0.37545  1.04103   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption)  \
122                   0.00000  0.40840                       

### Exercise 3 - sorting in descending order
---

Sort the data by Freedom and Trust (Government Corruption) in descending order and show the Country and Region only for the last five rows

**Test output**:
136, 117, 95, 101, 111 Country and Region columns


In [35]:
import pandas as pd
url = "https://github.com/futureCodersSE/working-with-data/blob/main/Happiness-Data/2015.xlsx?raw=true"
happiness = pd.read_excel(url)

sorted_Trust = happiness.sort_values (by= [ "Freedom", "Trust (Government Corruption)"],ascending = False).tail(5)
print (sorted_Trust )

                    Country                           Region  Happiness Rank  \
136                  Angola               Sub-Saharan Africa             137   
117                   Sudan               Sub-Saharan Africa             118   
95   Bosnia and Herzegovina       Central and Eastern Europe              96   
101                  Greece                   Western Europe             102   
111                    Iraq  Middle East and Northern Africa             112   

     Happiness Score  Standard Error  Economy (GDP per Capita)   Family  \
136            4.033         0.04758                   0.75778  0.86040   
117            4.550         0.06740                   0.52107  1.01404   
95             4.949         0.06913                   0.83223  0.91916   
101            4.857         0.05062                   1.15406  0.92933   
111            4.677         0.05232                   0.98549  0.81889   

     Health (Life Expectancy)  Freedom  Trust (Government Corruption

# Cleaning the data

Data comes from a range of sources:  forms, monitoring devices, etc.  There will often be missing values, duplicate records and values that are incorrectly formatted.  These can affect summary statistics and graphs plotted from the data.

Techniques for data cleansing include:
*  removing records with missing or null data (NaN, NA, "")
*  removing duplicate rows (keeping just one, either the first or the last)

Removal of rows according to criteria, or of columns are other ways that data might be cleaned up.  


---

## Removing NaN/Dropping NA values

pandas have functions for checking a dataframe, or column, for null values, checking a column for missing values, and functions for dropping all rows that contain null values.

* check for NA/NaN/missing values across dataframe (returns True if NA values exist)  
  `df.isnull().values.any()`  

* check for NA/NaN/missing values in specific column  
  `df["Make"].isnull().values.any()`  

* drop all rows that have NA/NaN values   
  `df.dropna()`  

* drop rows where NA/NaN values exist in specific columns  
  `df.dropna(subset = ["Make", "Model"])`  

### Exercise 4 - check for null values
---

1. read data from the file housing_in_london_yearly_variables.csv from this link: https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv
2. check if any NA values exist in the dataframe and print the result
3. use df.info() to see which columns have null entries (*Hint: if the non-null count is less than total entries, column contains missing/NA entries*)  

**Test output**:
True
.info shows median_salary, life_satisfaction, recycling_pct, population_size, number_of_jobs, area_size, no_of_houses all less than total rows (1071)



In [4]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

NA_result = df.isnull().values.any()
if NA_result == True:
  print ("some data is Null: ", NA_result)
else:
  print ("all dataframe has values", NA_result)

df.info()

some data is Null:  True
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1071 entries, 0 to 1070
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   code               1071 non-null   object 
 1   area               1071 non-null   object 
 2   date               1071 non-null   object 
 3   median_salary      1049 non-null   float64
 4   life_satisfaction  352 non-null    float64
 5   mean_salary        1071 non-null   object 
 6   recycling_pct      860 non-null    object 
 7   population_size    1018 non-null   float64
 8   number_of_jobs     931 non-null    float64
 9   area_size          666 non-null    float64
 10  no_of_houses       666 non-null    float64
 11  borough_flag       1071 non-null   int64  
dtypes: float64(6), int64(1), object(5)
memory usage: 100.5+ KB


### Exercise 5 - remove null values
---

1. remove rows with NA values for `life_satisfaction` (use [ ] even if only one column in list)
2. remove all NA values across whole dataframe

**Test output**:  
1.  Row count reduced to 352 rows
2.  Row count reduced to 267 rows

In [9]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

#remove rows with NA values for life_satisfaction (use [ ] even if only one column in list)
df.dropna (subset = ["life_satisfaction"])
print (df ["life_satisfaction"].count())

#remove all NA values across whole dataframe
df.dropna()
print (df.count())

352
code                 1071
area                 1071
date                 1071
median_salary        1049
life_satisfaction     352
mean_salary          1071
recycling_pct         860
population_size      1018
number_of_jobs        931
area_size             666
no_of_houses          666
borough_flag         1071
dtype: int64


## Dropping duplicates
---

* To remove duplicate rows based on duplication of values in all columns  
  `df.drop_duplicates()`  

* To remove rows that have duplicate entries in a specified column  
  `df.drop_duplicates(subset = ['Make'])`  

* To remove rows that have duplicate entries in multiple columns  
  `df.drop_duplicates(subset = ['Make', 'Model'])`

* Remove duplicate rows keeping the last instance rather than the first (default):  
  `df.drop_duplicates(keep='last')`  

### Exercise 6 - Removing duplicate entries
---

remove duplicate `area` entries keeping first instance  

**Test output**:  
 Dataframe now contains 50 rows all with date 1999-12-*01*

In [17]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

df.drop_duplicates (subset = ["area"])


code                      area        date  median_salary  \
0   E09000001            city of london  1999-12-01        33020.0   
1   E09000002      barking and dagenham  1999-12-01        21480.0   
2   E09000003                    barnet  1999-12-01        19568.0   
3   E09000004                    bexley  1999-12-01        18621.0   
4   E09000005                     brent  1999-12-01        18532.0   
5   E09000006                   bromley  1999-12-01        16720.0   
6   E09000007                    camden  1999-12-01        23677.0   
7   E09000008                   croydon  1999-12-01        19563.0   
8   E09000009                    ealing  1999-12-01        20580.0   
9   E09000010                   enfield  1999-12-01        19289.0   
10  E09000011                 greenwich  1999-12-01        21236.0   
11  E09000012                   hackney  1999-12-01        23249.0   
12  E09000013    hammersmith and fulham  1999-12-01        25000.0   
13  E09000014                  haringey  1999-12-01        18783.0   
14  E09000015                    harrow  1999-12-01        20596.0   
15  E09000016                  havering  1999-12-01        17165.0   
16  E09000017                hillingdon  1999-12-01        24002.0   
17  E09000018                  hounslow  1999-12-01        20155.0   
18  E09000019                 islington  1999-12-01        25113.0   
19  E09000020    kensington and chelsea  1999-12-01        20646.0   
20  E09000021      kingston upon thames  1999-12-01        19302.0   
21  E09000022                   lambeth  1999-12-01        23151.0   
22  E09000023                  lewisham  1999-12-01        20580.0   
23  E09000024                    merton  1999-12-01        18962.0   
24  E09000025                    newham  1999-12-01        18862.0   
25  E09000026                 redbridge  1999-12-01        19580.0   
26  E09000027      richmond upon thames  1999-12-01        22321.0   
27  E09000028                 southwark  1999-12-01        22784.0   
28  E09000029                    sutton  1999-12-01        19582.0   
29  E09000030             tower hamlets  1999-12-01        26376.0   
30  E09000031            waltham forest  1999-12-01        18547.0   
31  E09000032                wandsworth  1999-12-01        21321.0   
32  E09000033               westminster  1999-12-01        24447.0   
33  E12000001                north east  1999-12-01        16282.0   
34  E12000002                north west  1999-12-01        16977.0   
35  E12000003  yorkshire and the humber  1999-12-01        16527.0   
36  E12000004             east midlands  1999-12-01        16392.0   
37  E12000005             west midlands  1999-12-01        17000.0   
38  E12000006                      east  1999-12-01        18000.0   
39  E12000007                    london  1999-12-01        22487.0   
40  E12000008                south east  1999-12-01        18737.0   
41  E12000009                south west  1999-12-01        16727.0   
42  E13000001              inner london  1999-12-01            NaN   
43  E13000002              outer london  1999-12-01            NaN   
44  E92000001                   england  1999-12-01        17939.0   
45  K02000001            united kingdom  1999-12-01        17803.0   
46  K03000001             great britain  1999-12-01        17866.0   
47  K04000001         england and wales  1999-12-01        17974.0   
48  N92000002          northern ireland  1999-12-01        15798.0   
49  S92000003                  scotland  1999-12-01        16914.0   
50  W92000004                     wales  1999-12-01        16457.0   

    life_satisfaction mean_salary recycling_pct  population_size  \
0                 NaN       48922             0           6581.0   
1                 NaN       23620             3         162444.0   
2                 NaN       23128             8         313469.0   
3                 NaN       21386            18         217458.0   
4                 NaN       

# Normalising Data  
When we normalise data, we remodel a numeric column in a dataframe to be on a standard scale (e.g. 0 or 1).   

For example if we had a column of BMI scores, we could normalise that column so that all scores greater than or equal to 25 were recoded to the value 1 (bad) and all scores less than 25 were recoded to 0 (good).  

To normalise we need to:
*   write a function, with the dataframe as a parameter, which will look at each row in dataframe column and return either a value in the normalised scale (e.g. 0,1 or 1,2,3,4) depending on that value.

For example:  
```
def normalise_bmi(df):
  if df['bmi'] >= 25:
    return 1
  else:
    return 0

df["bmi"] = df.apply(normalise_bmi, axis=1)
```
This code reassigns the values in the column "bmi" by sending each row one after the other to the normalise_bmi function, which will check the value in the "bmi" column and return either 0 or 1 depending on the value in the "bmi" column.

### Exercise 7 - normalise data set
---

Create a function called **normalise_income(df)** that will return the values 1, 2 or 3 to represent low income, middle income and high income.  If the value in `df['median_salary']` is less than 27441 (the median), return 1, otherwise if it is less than 30932 (the upper quartile) return 2 and otherwise return 3.

Apply the normalise_income(df) function to the `median_salary` column.

*NOTE:  this operation will change the original dataframe so if you run it twice, everything in the median_salary column will change to 1 (as it had already been reduced to 1, 2 or 3 - if this happens, run the code in Exercise 4 again to get the original data again from the file.*

**Test output**:  
The maximum value of the column df['median_salary'] will be 3 and the minimum value will be 1  

In [19]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

def normalise_income(df):
  if df["median_salary"] <= 27441:
    return 1
  elif df ["median_salary"] <= 30932:
    return 2
  else:
    return 3

df ["median_salary"] = df.apply (normalise_income, axis=1)
print (df["median_salary"])

0       3
1       1
2       1
3       1
4       1
       ..
1066    2
1067    2
1068    1
1069    2
1070    2
Name: median_salary, Length: 1071, dtype: int64


### Exercise 8 - normalise the number of jobs column
---

Using what you have learnt from Exercise 7:  
*  use `df.describe()` to find the median, upper quartile and maximum for the number_of_jobs column  
*  create a function called **normalise_jobs(df)** that will return 1 if the `number_of_jobs` is below the median, 2 if the `number_of_jobs` is below the upper quartile or 3 otherwise.
*  normalise the `number_of_jobs` column by applying the function `normalise_jobs`.

**Test output**:  
The maximum value of the column df['number_of_jobs'] will be 3 and the minimum value will be 1  

In [25]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

df.describe()

def normalise_jobsNo (df):
  if df ["number_of_jobs"] <= 3188000:
    return 1
  elif df ["number_of_jobs"] <= 22170000:
    return 2
  else:
    return 3
df ["number_of_jobs"] = df.apply (normalise_jobsNo, axis=1)
print (df ["number_of_jobs"])

print (df ["number_of_jobs"].max(), df ["number_of_jobs"].min())

0       3
1       3
2       3
3       3
4       3
       ..
1066    3
1067    3
1068    3
1069    3
1070    3
Name: number_of_jobs, Length: 1071, dtype: int64
3 1


## Exercise 9 - normalise into a new column
---

Create a new function and code to normalise the `no_of_houses` column BUT this time, instead of assigning the result to `df['no_of_houses']` assign it to a new column called `df['housing_volume']`

**Test output**:  
The maximum value of the column df['housing_volume'] will be 3 and the minimum value will be 1

In [29]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

df.describe()

def nornalize_houses (df):
  if df ["no_of_houses"] <= 8814000:
    return 1
  elif df ["no_of_houses"] <= 12620000:
    return 2
  else:
    return 3
df ["housing_volume"] = df.apply (normalise_jobsNo, axis=1)
print (df ["housing_volume"])

print (df ["housing_volume"].max(), df ["housing_volume"].min())


0       3
1       3
2       3
3       3
4       3
       ..
1066    3
1067    3
1068    3
1069    3
1070    3
Name: housing_volume, Length: 1071, dtype: int64
3 1
           code                      area        date  median_salary  \
1021  E09000002      barking and dagenham  2019-12-01        28738.0   
1022  E09000003                    barnet  2019-12-01        31624.0   
1023  E09000004                    bexley  2019-12-01        31642.0   
1024  E09000005                     brent  2019-12-01        31739.0   
1025  E09000006                   bromley  2019-12-01            NaN   
1026  E09000007                    camden  2019-12-01        40196.0   
1027  E09000008                   croydon  2019-12-01        35237.0   
1028  E09000009                    ealing  2019-12-01        31021.0   
1029  E09000010                   enfield  2019-12-01        28258.0   
1030  E09000011                 greenwich  2019-12-01        33081.0   
1031  E09000012                   hackney  20

### Exercise 10 - normalise boroughs
---

Normalise the `area_size` column so that all values below mean are represented as 0 and otherwise are 1.  Assign the output to a new column called `area_size_normalised`.  

**Test output**:  
`area_size_normalised` column will contain both 0s and 1s.  The position of the first row with value 1 will be 0 and the position of the first row with value 0 will be 102.


In [4]:
import pandas as pd
url = "https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv"
df = pd.read_csv(url)

df.describe()
df.dropna (subset = ["area_size"])

def normalize_area (df):
  if df ["area_size"] <= 3724000:
    return 0
  else:
    return 1
df ["area_size_normalised"] = df.apply (normalize_area, axis = 1)

print (df.head (10))



        code                  area        date  median_salary  \
0  E09000001        city of london  1999-12-01        33020.0   
1  E09000002  barking and dagenham  1999-12-01        21480.0   
2  E09000003                barnet  1999-12-01        19568.0   
3  E09000004                bexley  1999-12-01        18621.0   
4  E09000005                 brent  1999-12-01        18532.0   
5  E09000006               bromley  1999-12-01        16720.0   
6  E09000007                camden  1999-12-01        23677.0   
7  E09000008               croydon  1999-12-01        19563.0   
8  E09000009                ealing  1999-12-01        20580.0   
9  E09000010               enfield  1999-12-01        19289.0   

   life_satisfaction mean_salary recycling_pct  population_size  \
0                NaN       48922             0           6581.0   
1                NaN       23620             3         162444.0   
2                NaN       23128             8         313469.0   
3               